### PR TITLE
Add support writing large files/empty files

### PR DESCRIPTION
**Pull request by ajeddeloh**
Use io.Reader instead of reading whole files into memory. Prevent
ignition from using large amounts of memory when writing large files.

Also make default http timeout for connection, not the entire transfer

### CODE REVIEW

File: internal/util/http.go [hunk 1]
```diff
@@ -35,7 +36,10 @@ type HttpClient struct {
 func NewHttpClient(logger *log.Logger) HttpClient {
 	return HttpClient{
 		client: &http.Client{
-			Timeout: 10 * time.Second,
+			//Timeout: 10 * time.Second,
```
Comment (crawford) [inline internal/util/http.go]: Toss this line.

File: internal/util/http.go [hunk 2]
```diff
@@ -81,6 +85,48 @@ func (c HttpClient) GetWithHeader(url string, header http.Header) ([]byte, int,
 	return body, status, err
 }
 
+//TODO change this comment
```
Comment (crawford) [inline internal/util/http.go]: TODO

File: internal/util/http.go [hunk 3]
```diff
@@ -81,6 +85,48 @@ func (c HttpClient) GetWithHeader(url string, header http.Header) ([]byte, int,
...
+// and returns the response body, HTTP status code, and error (if any). By
+// default, User-Agent and Accept are added to the header but these can be
+// overridden.
+func (c HttpClient) GetStreamWithHeader(url string, header http.Header) (io.ReadCloser, int, error) {
+	var body io.ReadCloser
+	var status int
+
+	err := c.logger.LogOp(func() error {
+		req, err := http.NewRequest("GET", url, nil)
+		req.Header.Set("User-Agent", "Ignition/"+version.Raw)
```
Comment (crawford) [inline internal/util/http.go]: Can we pull these headers into a common location?

File: internal/util/url.go [hunk 4]
```diff
@@ -107,6 +110,96 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
...
+func (f UnmountCloser) Close() error {
+	if err := f.Closer.Close(); err != nil {
+		return err;
+	}
+	f.umount();
+	return nil
+}
+
+//caller must close reader
+func FetchResourceAsStream(l *log.Logger, u url.URL) (io.ReadCloser, error) {
```
Comment (crawford) [inline internal/util/url.go]: Let's not duplicate so much code.

**Comment by ajeddeloh** [comment]
Added HTTP timeouts, docs.

File: internal/exec/util/file.go [hunk 5]
```diff
@@ -34,61 +35,82 @@ const (
...
-	}
-
-	decompress := func() error {
-		contents, err = decompressFile(l, f, contents)
-		return err
-	}
-
-	if l.LogOp(fetch, "fetching file %q", f.Path) != nil {
+	var (
+		contents io.ReadCloser
```
Comment (crawford) [inline internal/exec/util/file.go]: var var var

File: internal/util/url.go [hunk 6]
```diff
@@ -45,18 +48,51 @@ const (
...
+type UnmountCloser struct {
+	umount func()
+	io.Closer
+}
+
+func (f UnmountCloser) Close() error {
+	if err := f.Closer.Close(); err != nil {
+		return err
+	}
+	f.umount()
```
Comment (crawford) [inline internal/util/url.go]: always unmount

File: internal/util/url.go [hunk 7]
```diff
@@ -45,18 +48,51 @@ const (
...
+	return nil
+}
+
+//caller must close reader
```
Comment (crawford) [inline internal/util/url.go]: clean comment

File: internal/exec/util/file.go [hunk 8]
```diff
@@ -34,61 +35,82 @@ const (
...
+		err      error
+		errFunc  func() error
+	)
+
+	contents, err = util.FetchResourceAsStream(l, url.URL(f.Contents.Source))
```
Comment (crawford) [inline internal/exec/util/file.go]: reader

File: config/types/timeouts.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+	HttpResponseHeaders int `json:"httpResponseHeaders,omitempty"`
+	HttpTotal           int `json:"httpTotal,omitempty"`
+}
+
+type timeouts Timeouts
+
+func (t *Timeouts) UnmarshalJSON(data []byte) error {
+	t.HttpResponseHeaders = defaultResponseHeaders
+	t.HttpTotal = defaultTotal
+	th := timeouts(*t)
```
Comment (crawford) [inline config/types/timeouts.go]: Use a temporary variable for `t` since you are making a copy anyway. No reason to change the Timeouts object.

File: internal/util/http.go [hunk 10]
```diff
@@ -25,6 +25,17 @@ import (
 	"github.com/coreos/ignition/internal/version"
 )
 
+// A timeout of 0 indicates no timeout
+var (
+	httpResponseHeaderTimeout int
```
Comment (crawford) [inline internal/util/http.go]: I really don't like global variables.

Comment (ajeddeloh) [inline internal/util/http.go]: What other options are there in this case? The timeouts are applied globally. The only other alternative I see is passing either them (or the Timeouts struct) to every function that might use make http requests, which also doesn't seem very clean

Comment (crawford) [inline internal/util/http.go]: Pass them as arguments to `NewHttpClient`. If we are calling that too many times, maybe we should reuse HttpClients a bit more.

File: config/types/timeouts.go (hunk 9 above)
Comment (ajeddeloh) [inline config/types/timeouts.go]: So it doesn't apply defaults if it exits due to error in json.UnmarshalJSON()? In this case don't we want the defaults anyway? Or is there some other reason I'm missing?

Comment (crawford) [inline config/types/timeouts.go]: > So it doesn't apply defaults if it exits due to error in json.UnmarshalJSON()?

Correct.

> In this case don't we want the defaults anyway?

No. We shouldn't modify the underlying object unless unmarshalling was successful.

> should the other types that work this way (e.g. filesystems, raid, etc) be changed as well?

Those types work the way I described. They don't need to change.

File: internal/util/http.go (hunk 10 above)
Comment (ajeddeloh) [inline internal/util/http.go]: AFAIK we don't reuse them at all and instead create a new one every time we make a request (which is not great). We could switch to something like GetHttpClient, which would hand you either a new one (if not created yet) or one that has already been created, but we would still need a global to store the HttpClient.

Comment (crawford) [inline internal/util/http.go]: Should probably just reuse them then.

**Comment by ajeddeloh** [comment]
Bump, moved the handling of defaults from the config/types to internal/util/http so config can remain a pure representation of the json its passed. Also switched to reusing http clients.

File: internal/exec/util/file.go [hunk 11]
```diff
@@ -33,62 +36,108 @@ const (
...
+	hasher, err := util.GetHasher(f.Contents.Verification)
+	if err != nil {
+		l.Crit("Error verifying file %q: %v", f.Path, err)
+		return nil
 	}
 
-	decompress := func() error {
-		contents, err = decompressFile(l, f, contents)
-		return err
+	if hasher != nil {
```
Comment (marineam) [inline internal/exec/util/file.go]: This can be nil despite GetHasher reporting no error?

Comment (ajeddeloh) [inline internal/exec/util/file.go]: Yes, because verification is optional. I wasn't sure whether it would be best to return a nil or something like a ioutil.Discard

File: internal/exec/util/file.go [hunk 12]
```diff
@@ -33,62 +36,108 @@ const (
...
-	Mode     os.FileMode
-	Uid      int
-	Gid      int
+	io.ReadCloser
+	hash.Hash
+	Path types.Path
+	Mode os.FileMode
+	Uid  int
+	Gid  int
+	hash string
```
Comment (marineam) [inline internal/exec/util/file.go]: Perhaps name this expectedhash or some thing else more descriptive, otherwise it isn't clear what it is compared to the embedded Hash

File: internal/util/verification.go [hunk 13]
```diff
@@ -47,11 +48,25 @@ func AssertValid(verify types.Verification, data []byte) error {
...
 			}
 		}
 	}
 
 	return nil
 }
+
+func GetHasher(verify types.Verification) (hash.Hash, error) {
+	hash := verify.Hash
+	if hash == nil {
```
Comment (marineam) [inline internal/util/verification.go]: just use `verify.Hash == nil` and `verify.Hash.Function` below. Not used enough that a local var is helpful and you are masking the package named `hash` which can be confusing.

File: internal/util/url.go [hunk 14]
```diff
@@ -80,27 +114,40 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
...
 		} else if err != nil {
 			l.Err("failed to read oem config: %v", err)
 			return nil, ErrFailed
 		}
 
-		return data, nil
+		return struct {
+			io.Reader
+			io.Closer
+		}{bufio.NewReader(f), closer}, nil
```
Comment (marineam) [inline internal/util/url.go]: What is the point of using bufio here? If you are only exposing a ReadCloser you aren't doing anything like pushing input back into the buffer or whatnot.

File: internal/util/url.go [hunk 15]
```diff
@@ -80,27 +114,40 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
 
 		// check if present under oemDirPath, if so use it.
 		absPath := filepath.Join(oemDirPath, path)
-		data, err := ioutil.ReadFile(absPath)
+		f, err := os.Open(absPath)
+		var closer io.Closer = f
```
Comment (marineam) [inline internal/util/url.go]: why do you need this `closer` variable? Just use f

Comment (marineam) [inline internal/util/url.go]: oh, I see, you are using a single return. But if you just flip the conditionals you can put `if err == nil { return ... } else if !os.IsExistErr(err) { return err... }` at the top and remove one level of nesting from the bulk of the code.

File: internal/util/url.go (hunk 14 above)
Comment (ajeddeloh) [inline internal/util/url.go]: It's reading from a file; don't you usually want to buffer that?

File: internal/util/url.go [hunk 16]
```diff
@@ -45,18 +48,50 @@ const (
...
+	return data, nil
+}
+
+// Calls umount() after close
+type UnmountCloser struct {
+	umount func()
+	io.Closer
+}
+
+func (f UnmountCloser) Close() error {
```
Comment (marineam) [inline internal/util/url.go]: If you stored the logger instead of a function usage would be a little simpler to follow and close would be:

```go
func (f UnmountCloser) Close() error {
    defer umountOEM(f.logger)
    return f.Closer.Close()
}
```

Or perhaps call it a ReadUnmounter so you implement ReadCloser and avoid wrapping the underlying file in yet another struct below.

File: internal/util/url.go (hunk 14 above)
Comment (marineam) [inline internal/util/url.go]: Usually no

File: internal/exec/util/file.go [hunk 17]
```diff
@@ -33,62 +36,108 @@ const (
...
+		l.Crit("Error verifying file %q: %v", f.Path, err)
+		return nil
 	}
 
-	decompress := func() error {
-		contents, err = decompressFile(l, f, contents)
-		return err
+	if hasher != nil {
+		hash = f.Contents.Verification.Hash.Sum
+		reader = struct {
```
Comment (marineam) [inline internal/exec/util/file.go]: How about put this in a NewHashedReader function so intent is clear, defining things with anonymous structs inline with real code clutters and distracts a bit.

**Comment by ajeddeloh** [comment]
Updated to only have code for conversion to using io.Reader and writing empty files. Switched to using timeout for getting headers instead of whole file when using http. No longer contains code for making timeouts configurable.

File: internal/exec/util/file.go [hunk 18]
```diff
@@ -33,69 +36,123 @@ const (
...
+	if string(encodedSum) != f.expectedHash {
+		return util.ErrHashMismatch{
+			Calculated: string(encodedSum),
+			Expected:   f.expectedHash,
+		}
+	}
+	return nil
+}
+
+//returns a new ReadCloser that also writes to the provided hasher
```
Comment (crawford) [inline internal/exec/util/file.go]: Comments should start with the function name.

File: internal/exec/util/file.go [hunk 19]
```diff
@@ -33,69 +36,123 @@ const (
...
+}
+
+// Returns a *File with a Reader that downloads, sha512's, and unzips the incoming data
+// Returns nil if f had invalid options. Errors reading/verifying/uncompressing the file will
+// present themselves when the Reader is actually read from
 func RenderFile(l *log.Logger, f types.File) *File {
-	var contents []byte
+	var reader io.ReadCloser
 	var err error
+	var expectedHash string
```
Comment (crawford) [inline internal/exec/util/file.go]: Why is this needed?

File: internal/util/url.go [hunk 20]
```diff
@@ -80,27 +113,46 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
...
 			return nil, ErrFailed
 		}
 
-		return data, nil
+		return ReadUnmounter{
+			logger:     l,
+			ReadCloser: f,
+		}, nil
+
+	case "":
```
Comment (crawford) [inline internal/util/url.go]: Break this into a separate commit.

File: internal/util/url.go [hunk 21]
```diff
@@ -80,27 +113,46 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
...
-				l.Err("failed to mount oem partition: %v", err)
-				return nil, ErrFailed
-			}
-
-			absPath := filepath.Join(oemMountPath, path)
-			data, err = ioutil.ReadFile(absPath)
-			umountOEM(l)
-		} else if err != nil {
+		f, err := os.Open(absPath)
+
```
Comment (crawford) [inline internal/util/url.go]: Remove newline

File: internal/exec/util/file.go [hunk 22]
```diff
@@ -33,69 +36,123 @@ const (
...
+	}
+	return nil
+}
+
+//returns a new ReadCloser that also writes to the provided hasher
+func newHashedReader(reader io.ReadCloser, hasher hash.Hash) io.ReadCloser {
+	return struct {
+		io.Reader
+		io.Closer
+	}{io.TeeReader(reader, hasher), reader}
```
Comment (crawford) [inline internal/exec/util/file.go]: Use explicit field assignment

File: internal/exec/util/file.go (hunk 19 above)
Comment (crawford) [inline internal/exec/util/file.go]: Nevermind

File: internal/exec/util/file.go [hunk 23]
```diff
@@ -33,69 +36,123 @@ const (
...
+		l.Crit("Error decompressing file %q: %v", f.Path, err)
 		return nil
 	}
-	if l.LogOp(validate, "validating file contents") != nil {
-		return nil
+
+	return &File{
+		Path:         f.Path,
+		ReadCloser:   reader,
+		Hash:         hasher,
```
Comment (crawford) [inline internal/exec/util/file.go]: Can you clean up these names a bit? Assigning `hasher` to `Hash` is odd.

File: internal/exec/util/file.go [hunk 24]
```diff
@@ -33,69 +36,123 @@ const (
...
-	Mode     os.FileMode
-	Uid      int
-	Gid      int
+	io.ReadCloser
+	hash.Hash
+	Path         types.Path
+	Mode         os.FileMode
+	Uid          int
+	Gid          int
+	expectedHash string
```
Comment (crawford) [inline internal/exec/util/file.go]: expectedSum

File: internal/exec/util/file.go [hunk 25]
```diff
@@ -33,69 +36,123 @@ const (
...
+		Uid:          f.User.Id,
+		Gid:          f.Group.Id,
+		expectedHash: expectedHash,
 	}
-	if l.LogOp(decompress, "decompressing file contents") != nil {
-		return nil
+}
+
+// Wrapper for gzip's reader that closes the stream it wraps as well
+type gzipReader struct {
```
Comment (crawford) [inline internal/exec/util/file.go]: Is there a better pattern for this in Go? This feels wrong.

File: internal/util/http.go [hunk 26]
```diff
@@ -24,19 +25,49 @@ import (
 	"github.com/coreos/ignition/internal/version"
 )
 
+const (
+	defaultResponseHeadersTimeout int = 10
+	defaultTotalTimeout           int = 0
+)
+
+// Http client to be reused
+var sharedClient *http.Client
```
Comment (crawford) [inline internal/util/http.go]: Still don't like globals.

File: internal/util/http.go [hunk 27]
```diff
@@ -24,19 +25,49 @@ import (
 	"github.com/coreos/ignition/internal/version"
 )
 
+const (
+	defaultResponseHeadersTimeout int = 10
```
Comment (crawford) [inline internal/util/http.go]: Break this into a different commit.

File: internal/exec/util/file.go (hunk 25 above)
Comment (ajeddeloh) [inline internal/exec/util/file.go]: The problem here is that gzip.NewReader() takes in a io.Reader but returns an io.ReadCloser (interestingly, it implements the io.ReadCloser interface, but the type signature of the function says it returns a io.Reader). Close() on the gzip.Reader does not actually close reader the it is wrapping, which is why I have the extra type that Close()'s both.

Comment (crawford) [inline internal/exec/util/file.go]: > but the type signature of the function says it returns a io.Reader

Implicit interface implementation is such a great feature.

> Close() on the gzip.Reader does not actually close reader the it is wrapping

Yeah, I noticed. What a pain in the ass.

**Comment by ajeddeloh** [comment]
bump with changes sugguested

File: internal/util/http.go [hunk 28]
```diff
@@ -32,11 +32,13 @@ type HttpClient struct {
 	logger *log.Logger
 }
 
-// NewHttpClient creates a new client with the given logger.
```
Comment (crawford) [inline internal/util/http.go]: Comments should be English sentences. Leave the period.

File: internal/util/http.go [hunk 29]
```diff
@@ -36,7 +36,9 @@ type HttpClient struct {
 func NewHttpClient(logger *log.Logger) HttpClient {
 	return HttpClient{
 		client: &http.Client{
-			Timeout: 10 * time.Second,
+			Transport: &http.Transport{
```
Comment (dghubble) [inline internal/util/http.go]: From the net/http package,

> // If nil, DefaultTransport is used.
        Transport RoundTripper

So 524677a replaces the DefaultTransport.

Comment (ajeddeloh) [inline internal/util/http.go]: Fixed. I though empty value for Transport would be the same as DefaultTransport, but it turns out its not. Copied the default values, expect the ones which don't apply or aren't in go 1.5.

**Comment by crawford** [comment]
Please clean up comments according to https://golang.org/doc/effective_go.html#commentary.

File: internal/exec/util/file.go [hunk 30]
```diff
@@ -109,30 +170,38 @@ func (u Util) WriteFile(f *File) error {
...
+	if err = f.Verify(); err != nil {
 		return err
 	}
 
 	// XXX(vc): Note that we assume to be operating on the file we just wrote, this is only guaranteed
 	// by using syscall.Fchown() and syscall.Fchmod()
 
 	// Ensure the ownership and mode are as requested (since WriteFile can be affected by sticky bit)
-	if err := os.Chown(tmp.Name(), f.Uid, f.Gid); err != nil {
+	if err = os.Chown(tmp.Name(), f.Uid, f.Gid); err != nil {
```
Comment (crawford) [inline internal/exec/util/file.go]: Break these into a separate commit.

File: internal/util/url.go [hunk 31]
```diff
@@ -45,18 +47,50 @@ const (
 
 // FetchResource fetches a resource given a URL. The supported schemes are http, data, and oem.
 func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
+	var data []byte
+
+	dataReader, err := FetchResourceAsReader(l, u)
+
```
Comment (crawford) [inline internal/util/url.go]: remove line

File: internal/util/url.go [hunk 32]
```diff
@@ -45,18 +47,50 @@ const (
...
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (crawford) [inline internal/util/url.go]: remove line

File: internal/util/url.go [hunk 33]
```diff
@@ -80,27 +113,37 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
...
-			if err != nil {
-				l.Err("failed to mount oem partition: %v", err)
-				return nil, ErrFailed
-			}
-
-			absPath := filepath.Join(oemMountPath, path)
-			data, err = ioutil.ReadFile(absPath)
-			umountOEM(l)
-		} else if err != nil {
+		f, err := os.Open(absPath)
```
Comment (crawford) [inline internal/util/url.go]:
```go
if f, err := os.Open(absPath); err == nil {
    return f, nil
} else if !os.IsNotExist(err) {
```

**Comment by ajeddeloh** [comment]
Updated comments, fixed linebreaks, broke error handling fix into seperate commit.

**Comment by crawford** [comment]
LGTM. Please rebase.
